std/experimental/allocator/building_blocks/free_list.d

```diff
@@ -0,0 +1,1034 @@
+module std.experimental.allocator.building_blocks.free_list;
+
+import std.experimental.allocator.common;
+import std.typecons : Flag, Yes, No;
+
+/**
+
+$(WEB en.wikipedia.org/wiki/Free_list, Free list allocator), stackable on top of
+another allocator. Allocation requests between $(D min) and $(D max) bytes are
+rounded up to $(D max) and served from a singly-linked list of buffers
+deallocated in the past. All other allocations are directed to $(D
+ParentAllocator). Due to the simplicity of free list management, allocations
+from the free list are fast.
+
+One instantiation is of particular interest: $(D FreeList!(0, unbounded)) puts
+every deallocation in the freelist, and subsequently serves any allocation from
+the freelist (if not empty). There is no checking of size matching, which would
+be incorrect for a freestanding allocator but is both correct and fast when an
+owning allocator on top of the free list allocator (such as $(D Segregator)) is
+already in charge of handling size checking.
+
+The following methods are defined if $(D ParentAllocator) defines them, and
+forward to it: $(D expand), $(D owns), $(D reallocate).
+
+*/
+struct FreeList(ParentAllocator,
```

Comment:
FreeList has no destructor. This means that if there is a free list backed by Mallocator, any memory that user code tried to deallocate but was still on the free list at the time that the list went out of scope will be leaked.